### PR TITLE
Create separate OpSource for each included file when compiling -g

### LINKED ISSUE #1615: Need separate OpSource for each included file for SPIR-V -g compilation
This is a replacement issue for #1611. That issue contains the rationale for this proposal.

This issue proposes that for each different source file that is #include'd during -g compilation of a SPIR-V module, a new OpSource instruction is inserted in the SPIR-V which contains the original source of the included file.

### PR DESCRIPTION
**Pull request by greg-lunarg**
Fixes #1615

### CODE REVIEW

File: SPIRV/SpvBuilder.h [hunk 1]
```diff
@@ -74,20 +75,33 @@ class Builder {
...
     }
     void setSourceText(const std::string& text) { sourceText = text; }
     void addSourceExtension(const char* ext) { sourceExtensions.push_back(ext); }
     void addModuleProcessed(const std::string& p) { moduleProcesses.push_back(p.c_str()); }
     void setEmitOpLines() { emitOpLines = true; }
     void addExtension(const char* ext) { extensions.insert(ext); }
+    void addInclude(const std::string& name, const std::string& text)
+    {
+        spv::Id incId = getStringId(name);
+        includeFiles[incId] = text;
```
Comment (johnkslang) [inline SPIRV/SpvBuilder.h]: Makes a copy.

File: SPIRV/GlslangToSpv.cpp [hunk 2]
```diff
@@ -1297,6 +1297,10 @@ TGlslangToSpvTraverser::TGlslangToSpvTraverser(unsigned int spvVersion, const gl
             text.append("#line 1\n");
         text.append(glslangIntermediate->getSourceText());
         builder.setSourceText(text);
+        // Pass name and text for all included files
+        const std::map<std::string, std::string>& include_txt = glslangIntermediate->getIncludeText();
+        for (auto& inc : include_txt)
+            builder.addInclude(inc.first, inc.second);
```
Comment (johnkslang) [inline SPIRV/GlslangToSpv.cpp]: This just makes a mirror copy of everything?  The AST survives through the whole process.

File: SPIRV/SpvBuilder.h [hunk 3]
```diff
@@ -708,6 +723,9 @@ class Builder {
     // map from strings to their string ids
     std::unordered_map<std::string, spv::Id> stringIds;
 
+    // map from include file name ids to their contents
+    std::map<spv::Id, std::string> includeFiles;
```
[the remark this answers is not in the document]
Comment (johnkslang) [reply SPIRV/SpvBuilder.h]: I think the main copy can be in the AST and this map can point there, either a `string*` or a `const char*`, which ever is smoothest.

File: glslang/MachineIndependent/preprocessor/Pp.cpp [hunk 4]
```diff
@@ -653,6 +653,8 @@ int TPpContext::CPPinclude(TPpToken* ppToken)
             epilogue << (res->headerData[res->headerLength - 1] == '\n'? "" : "\n") <<
                 "#line " << directiveLoc.line + forNextLine << " " << directiveLoc.getStringNameOrNum() << "\n";
             pushInput(new TokenizableIncludeFile(directiveLoc, prologue.str(), res, epilogue.str(), this));
+            parseContext.intermediate.addIncludeText(res->headerName.c_str(),
+                std::string(res->headerData, res->headerLength).c_str());
```
[the remark this answers is not in the document]
Comment (johnkslang) [reply glslang/MachineIndependent/preprocessor/Pp.cpp]: I suggest having `addIncludeText` take the `char*, length` form and skipping this temporary copy in the calling argument.

**Comment by greg-lunarg** [comment]
@johnkslang I think I have fixed all your initial concerns.

**Pull request merged**